---
---
# EXPLORATORY DATA ANALYSIS

---
---

## ROADMAP: where we've been, and where we're going

In recent lectures we learned about **PPDAC**, a structured framework for data-driven problem solving:

<div style="text-align: center;">
<img src="https://raw.githubusercontent.com/ntstevens/DATSC101/58fc1f744c0caf95ae25e32836bb8625c59a40b7/images/PPDAC.svg" alt="PPDAC" width="800"/>
</div>

* **Problem:** a clear statement of the study's objectives, usually involving one or more questions.
* **Plan:** the procedures used to carry out the study including how we will collect the data.
* **Data:** the physical collection of the data, as described in the Plan.
* **Analysis:** the analysis of the data collected in light of the Problem and the Plan.
* **Conclusion:** The conclusions that are drawn from the Analysis about the Problem and their limitations.

In the coming lectures we begin to use this framework, stage by stage, with actual data and actual code. 

We begin with **exploratory data analysis** which is used to answer descriptive and exploratory questions by investigating datasets to summarize their main characteristics, uncover underlying relationships, and spot anomalies, primarily using statistical summaries and data visualization.


---
---

## About the Data: MLB Salaries

<div style="text-align: center;">
  <img src="https://img.mlbstatic.com/mlb-images/image/private/t_16x9/t_w2208/mlb/crbela7xbocusesq8f8v.jpg" alt="mlb_logo" width = 600/>
</div>

[Image Source](https://www.mlb.com/official-information)

Throughout this section we will illustrate ideas, principles, and tasks through the exploratory analysis of Major League Baseball (MLB) salaries.

The data we'll analyze is stored in the `mlb_2025_salaries.csv` file made available on both LEARN and [here on GitHub](https://github.com/ntstevens/DATSC101/blob/main/data/mlb_2025_salaries.csv). The data in its raw form comes from [spotrac.com](https://www.spotrac.com/mlb/rankings/player/_/year/2025/sort/cap_base), an online financial resource for professional sports.

The dataset is composed of salaries and other information for each of the 1512 players in the MLB during the 2025 season. For each player, the following information is recorded:

* `Player Name`: The player's name
* `Team Abbreviation`: The player's team, recorded as a 2 or 3-character abbreviation. A dictionary for the abbreviations is found [here](https://mlb10theshow.fandom.com/wiki/Team_Abbreviations).
* `Position`: The player's position(s), recorded as a 1 or 2-character abbreviations. A dictionary for the abbreviations is found [here](https://www.yourdictionary.com/articles/baseball-positions-abbreviations).
* `2025 Base Salary ($)`: The player's base salary for the 2025 season (in USD).


Let's load it in using the `pd.read_csv()` function we've used previously:

In [69]:
import pandas as pd

mlb_salaries = pd.read_csv('https://raw.githubusercontent.com/ntstevens/DATSC101/refs/heads/main/data/mlb_2025_salaries.csv')
mlb_salaries

,Player Name,Team Abbreviation,Position,2025 Base Salary ($)
0,A.J. Puk,ARI,RP/CL,2950000
1,Adrian Del Castillo,ARI,C,760000
2,Alek Thomas,ARI,CF,793000
3,Andrew Hoffman,ARI,RP,760000
4,Andrew Saalfrank,ARI,RP,760000
...,...,...,...,...
1507,Sauryn Lao,WSH,RP,760000
1508,Shinnosuke Ogasawara,WSH,RP,1500000
1509,Trevor Williams,WSH,SP,7000000
1510,Trey Lipscomb,WSH,3B,760000


### ACTIVITY 🗣

**Before we analyze this data, let's give some thought to the Problem, Plan, and Data.**

* What is a *unit* in this dataset?
* What is the *response variable* and what are the *explanatory variables*?
* What is the *target population*?
* What type of data is this?
* What are some questions we can answer with this data?
* What are some questions we can't answer with this data?


---
---

## Data Frames and Other Data Structures

In Tutorial 2 we introduced **tabular data**, data that can be stored in a spreadsheet in row-column format. Recall, tabular data is structured such that

* Rows represent the objects we study. We call these **observations**.
* Columns represent characteristics of those objects. We call these **variables**.
* A **cell** contains the value of a particular variable for a particular observation.

When we load tabular data into Python using `pandas`, it is represented as a **data frame** object. In this section we will dig deeper into what a data frame object is, and along the way consider other _data structures_.

Formally, a data frame is a collection of **series** stuck together. What is a series then?

* A series is an object that contains one or more elements of any _data type_ in a single column.
* We can create a series in `pandas` using `pd.Series()`:


In [5]:
import pandas as pd

courses = pd.Series(["MATH135", "MATH137", "CS125", "PSYCH101","MUSIC140"])
courses

0     MATH135
1     MATH137
2       CS125
3    PSYCH101
4    MUSIC140
dtype: object

As you can see above, this series has 5 elements and is composed of strings. But as noted above, a series can elements of _any_ data type. Here is another example in which the series is composed of numbers.

In [61]:
grades = pd.Series([92, 81, 80, 91, 87])
grades

0    92
1    81
2    80
3    91
4    87
dtype: int64

Note that the argument passed into `pd.Series()` above is called a **list**, which is a data structure composed of lists of elements, or potentially different data types inside square  brackets `[]`. Below is a summary of different data types you may encounter in Python. 

|Data Type | Abbreviation | Description | Example |
| --- | --- | --- | --- |
|integer | `int` | positive/negative/zero whole numbers | `101` |
|floating point number | `float` | decimal number | `3.14159` |
|boolean | `bool` | true or false | `True` |
|string | `str` | text | `"DATSC101"` |
|none | `NoneType` | no value | `None` |

Acknowledgement: this table is reproduced from a similar table given in our textbook [Data Science: A First Introduction with Python](https://python.datasciencebook.ca/wrangling.html#tab-datatype-table).

Although a single series can contain elements of multiple data types simultaneously (e.g., `pd.Series(["DATSC", 101])` which combines a string and and integer is technically permissible), we generally avoid it. 

This is because a series is typicaly used to represent a column in a data frame, and a column in a data frame records measurements on a single variable, and that variable prescribes a particular data type.

In fact, we can construct a data frame using multiple series. Below, we construct a 2-column data frame using the series `courses` and `grades` defined above.

In [65]:
df = pd.DataFrame({"Course": courses, "Grade": grades})
df

,Course,Grade
0,MATH135,92
1,MATH137,81
2,CS125,80
3,PSYCH101,91
4,MUSIC140,87


Note that the argument passed into `pd.DataFrame()` above is called a **dictionary**, which is a data structure composed of pairs of **keys** and **values** inside curly brackets `{}`. Each element starts with the `key` on the left, followed by a colon symbol `:`, and then the `value` on the right. A dictionary can have multiple key-value pairs, each separted by a comma, as we saw above.

Below is a summary of the various data structures we've encountered so far.

|Data Structure | Python Name | Description|
| --- | --- | --- |
|List | `list` | An ordered collection of elements of any data type. |
|Dictionary | `dict` | A labeled data structure where `keys` are paired with `values`. |
|Series | `Series` | An ordered and labelled collection of elements of any data type. | 
|Data frame | `DataFrame` | A labeled data structure with `Series` columns of potentially different types. |

Acknowledgement: this table is reproduced from a similar table given in our textbook [Data Science: A First Introduction with Python](https://python.datasciencebook.ca/wrangling.html#tab-datastruc-table).

While we work primarily with data frames, it's helpful to understand series and dictionaries because, as we've now seen, a data frame is composed of these. Next we look at working with a data frame, once we've read it in.

---
---

## Working with Data Frames

In this section we will overview how to:

* find helpful information about a data frame,
* subset the data frame by extracting specific rows and/or columns,
* modify rows and/or columns, and
* add rows and/or columns.

We illustrate all of this in the context of the MLB salary data.

### Learning about a data frame

Data frames have many important attributes that are helpful to know when understanding the data they contain. For instance, the number of rows and columns, the column names, and the data types stored in those columns.

We can access this information using built in **methods** associated with the `DataFrame` object itself. In particular, we will illustrate the use of the following:
* `df.shape`
* `df.columns`
* `df.rename`
* `df.info()`

Note that a _method_ is slightly different from a _function_. We've seen an used functions aleady. Recall their definition:

> *"A function is a reusable, self-contained block of code designed to perform a specific task."*

Previously, we passed an argument into a function, and that function performed some calculation on that input.

A method behaves similarly--it often performs some calculation--but its defined and applied explicitly with respect to some object. For our purposes, you can think of the difference simply in how the calculation is invoked.

For example, imagine some hypothetical calculation called `calculate()` we want to perform on an `object`. 

* The invocation of the *function* would be `calculate(object)`.
* The invocation of the *method* would be `object.calculate()`.

It turns out that when the `object` is a `DataFrame`, there are several helpful methods we can apply. The three above are just the tip of the iceburg. 


#### How many rows and columns does a data frame have?

Knowing the answer to this question tells you how many observations (rows) and variables (columns) are in the datset. We can use the `df.shape` method to determine this:

In [78]:
mlb_salaries.shape

(1512, 4)

We can see from the output above that there are 1512 rows and 4 columns. This agrees with the description of the data provided above: we have 4 variables (name, team, position, salary) measured on each of 1512 baseball players.

#### What are the column names? And can I change them?

We've seen that the header row in a data frame provides columns names. In small data frames this information might be obtained immediately by opening and viewing a `.csv` file. But in many real-world applications, there may be dozens, hundreds, or even thousands of columns, in which case have a simple line of code to view and store these names can be helpful. We use the `df.columns` method to do this.

In [168]:
mlb_salaries.columns

Index(['Name', 'Team', 'Position', 'Salary'], dtype='object')

Some of these column names are a bit cumbersome, especially if we have to type them again and again in the code the performs our analyses. In a case like this, it can be useful to relabel the columns with shorter (but still informative) names. We use the `df.rename` method to do this.

In [81]:
mlb_salaries.rename(columns={"Player Name": "Name", 
                             "Team Abbreviation": "Team",
                             "Position": "Position",
                             "2025 Base Salary ($)": "Salary"})

,Name,Team,Position,Salary
0,A.J. Puk,ARI,RP/CL,2950000
1,Adrian Del Castillo,ARI,C,760000
2,Alek Thomas,ARI,CF,793000
3,Andrew Hoffman,ARI,RP,760000
4,Andrew Saalfrank,ARI,RP,760000
...,...,...,...,...
1507,Sauryn Lao,WSH,RP,760000
1508,Shinnosuke Ogasawara,WSH,RP,1500000
1509,Trevor Williams,WSH,SP,7000000
1510,Trey Lipscomb,WSH,3B,760000


You can see above that we passed a dictionary into the `df.rename`, and in doing so we updated all of the column names except `Position`, which was already short and informative.

However, while the line of code above created a new data frame with updated column names, we didn't actually change the original `mlb_salaries` data frame. Observe:

In [82]:
mlb_salaries.columns

Index(['Player Name', 'Team Abbreviation', 'Position', '2025 Base Salary ($)'], dtype='object')

If we want to update the data frame itself, we have to overwrite the old one with the new one as follows:

In [84]:
mlb_salaries = mlb_salaries.rename(columns={"Player Name": "Name", 
                                            "Team Abbreviation": "Team",
                                            "Position": "Position",
                                            "2025 Base Salary ($)": "Salary"})
mlb_salaries.columns

Index(['Name', 'Team', 'Position', 'Salary'], dtype='object')

#### Is there an all-in-one method that gives me all this information?

The answer is yes, and it's the `df.info()` method. It tells you how many rows and columns the data frame contains, it tells you the column names, and it tells you the type of data contained in those columns.

In [169]:
mlb_salaries.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1512 entries, 0 to 1511
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Name      1512 non-null   object
 1   Team      1512 non-null   object
 2   Position  1512 non-null   object
 3   Salary    1512 non-null   int64 
dtypes: int64(1), object(3)
memory usage: 47.4+ KB


---
---

### Filtering Columns

In many data analyses we care about some columns more than others. In cases like these, we may choose to _filter_ our data frame to restrict attention to only the column(s) we're primariliy interested in. This is done with the square brackets `[]` operator. Within these square brackets, we need only specify the column(s) we wish to retain.

#### Filtering One Column

In the context of MLB salaries, we care most about the `Salary` column, so we could restrict attention to it alone.

In [170]:
mlb_salaries["Salary"]

0       2950000
1        760000
2        793000
3        760000
4        760000
         ...   
1507     760000
1508    1500000
1509    7000000
1510     760000
1511     760400
Name: Salary, Length: 1512, dtype: int64

This returns a `Series` of the 1512 salaries only. Note that the general principle works as follows:

<div style="text-align: center;">
  <img src="https://raw.githubusercontent.com/ntstevens/DATSC101/1d015e0ae312ec485e0f6d3e2e657a8933cbc8b7/images/column_filter.svg" alt="mlb_logo" height = "200"/>
</div>

#### Filtering Multiple Columns

In the context of MLB salaries, suppose we're most interested in how salaries change from one team to another, and we care less about positions. In this case, we could filter the columns such that `Position` is removed.

In [171]:
mlb_salaries[["Name", "Team", "Salary"]]

,Name,Team,Salary
0,A.J. Puk,ARI,2950000
1,Adrian Del Castillo,ARI,760000
2,Alek Thomas,ARI,793000
3,Andrew Hoffman,ARI,760000
4,Andrew Saalfrank,ARI,760000
...,...,...,...
1507,Sauryn Lao,WSH,760000
1508,Shinnosuke Ogasawara,WSH,1500000
1509,Trevor Williams,WSH,7000000
1510,Trey Lipscomb,WSH,760000


This returns a `DataFrame` with 3 columns. It is the original `mlb_salaries` data frame, but with the `Position` column removed. Note that the general principle works as follows:
<div style="text-align: center;">
  <img src="https://raw.githubusercontent.com/ntstevens/DATSC101/1d015e0ae312ec485e0f6d3e2e657a8933cbc8b7/images/columns_filter.svg" alt="mlb_logo" height = "200"/>
</div>

In both cases above, we filtered columns by combined the square bracket operator `[]` with column names. But it's important to notice that the number of brackets used depended on how many columns were being filtered:

* when filtering one column, we use one set of brackets `[]` and they contain the relevant column name
* when filtering multiple column, we use two sets of brackets `[[]]` and they contain the relevant column names, each separate by commas

**Aside:**

* In the latter example we simple dropped a column from the original data frame. We can equivalently use the `df.drop` method to remove one or more columns.

In [174]:
mlb_salaries.drop(columns = "Position")

,Name,Team,Salary
0,A.J. Puk,ARI,2950000
1,Adrian Del Castillo,ARI,760000
2,Alek Thomas,ARI,793000
3,Andrew Hoffman,ARI,760000
4,Andrew Saalfrank,ARI,760000
...,...,...,...
1507,Sauryn Lao,WSH,760000
1508,Shinnosuke Ogasawara,WSH,1500000
1509,Trevor Williams,WSH,7000000
1510,Trey Lipscomb,WSH,760000


---
---

### Filtering Rows

Filtering rows is as or more important in data analysis than filtering columns; filtering rows allows us to focus our analysis on particular subsets of the observations. It turns out we can also use the square bracket operator `[]` to filter a data frame's rows, but we use it differently than we did when filtering columns.

Because rows don't have names, we instead pass a _logical statement_ into `[]`. A logical statement is one that evaluates to `True` or `False` and arises through a logical comparison. The table below gives examples of logical comparisons as well as the corresponding Python syntax.

|Comparison | Operator | `True` example | `False` Example |
| --- | :-: | :-: | :-: |
|Less than | `<` | `2 < 3` | `2 < 2` |
|Greater than | `>` | `3 > 2` | `3 > 3` |
|Less than or equal | `<=` | `2 <= 2` | `3 <= 2` |
|Greater or equal | `>=` | `3 >= 3` | `2 >= 3` |
|Equal | `==` | `3 == 3` | `3 == 2` |
|Not equal | `!=` | `3 != 2` | `2 != 2` |

The goal is to establish a logical statement that specifies exactly which rows to include in the filtered data frame. Once such a statement is defined, we can filter the rows as follows:

<div style="text-align: center;">
  <img src="https://raw.githubusercontent.com/ntstevens/DATSC101/1d015e0ae312ec485e0f6d3e2e657a8933cbc8b7/images/row_filter.svg" alt="mlb_logo" height = "200"/>
</div>

In the context of MLB salaries, suppose we're most interested in the Toronto Blue Jays in which case we want to restrict attention to players whose `Team` is `TOR`. The following statement is a logical comparison checking whether each player's `Team` is `TOR`.

In [176]:
mlb_salaries["Team"] == "TOR"

0       False
1       False
2       False
3       False
4       False
        ...  
1507    False
1508    False
1509    False
1510    False
1511    False
Name: Team, Length: 1512, dtype: bool

The result is a `Series` of type `bool`. That is, a column of 1512 `True` and `False` values (one for each player). Although they're hidden in the suppressed output above, each of the elements corresponding to a Blue Jay is `True` and all other elements (corresponding to every play who is not a Blue Jay) are `False`.

With this logical statement defined, we're ready to filter `mlb_salaries`:

In [178]:
mlb_salaries[mlb_salaries["Team"] == "TOR"]

,Name,Team,Position,Salary
1417,Adam Macko,TOR,SP,123900
1418,Addison Barger,TOR,SS,760000
1419,Alejandro Kirk,TOR,C,4600000
1420,Andres Gimenez,TOR,2B,10000000
1421,Angel Bastardo,TOR,P,760000
1422,Anthony Santander,TOR,LF,13500000
1423,Bo Bichette,TOR,SS,16500000
1424,Bowden Francis,TOR,SP,781500
1425,Braydon Fisher,TOR,RP,760000
1426,Brendon Little,TOR,RP,771800


While the command above created the desired data frame, the new data frame is not saved. To do so, we need to assign it to a variable:

In [197]:
blue_jays = mlb_salaries[mlb_salaries["Team"] == "TOR"]

Note that because every player in this filtered data frame is a Blue Jay, the `Team` column is redundant. Let's remove it:

In [198]:
blue_jays = blue_jays.drop(columns = "Team")
blue_jays

,Name,Position,Salary
1417,Adam Macko,SP,123900
1418,Addison Barger,SS,760000
1419,Alejandro Kirk,C,4600000
1420,Andres Gimenez,2B,10000000
1421,Angel Bastardo,P,760000
1422,Anthony Santander,LF,13500000
1423,Bo Bichette,SS,16500000
1424,Bowden Francis,SP,781500
1425,Braydon Fisher,RP,760000
1426,Brendon Little,RP,771800


It would be interesting to sort this data frame to easily identify 

* who are the highest and lowest paid Blue Jays, and 
* what are those salaries? 

It turns out we have a method `df.sort_values` that does this for us. By default, it sorts from smallest to largest (`ascending = True`) but in this case it's more interesting to sort from largest to smallest (`ascending = False`).

In [200]:
blue_jays.sort_values(by = "Salary", ascending = False)

,Name,Position,Salary
1462,Vladimir Guerrero Jr.,1B,28500000
1443,Kevin Gausman,SP,23000000
1434,George Springer,CF,22500000
1427,Chris Bassitt,SP,21000000
1441,Jose Berrios,SP,18000000
1423,Bo Bichette,SS,16500000
1448,Max Scherzer,SP/SP2,15500000
1422,Anthony Santander,LF,13500000
1420,Andres Gimenez,2B,10000000
1457,Shane Bieber,SP/SP1,10000000


We find:
* the first basement Vladimir Guerrero Jr. has the largest base salary at \$28,500,000 and
* two starting pitchers, Jake Bloss and Adam Macko, have the lowest base salaries at \$123,900

Summing the values in the `Salary` column of the `blue_jays` data frame would give us the _team payroll_ (i.e., the total amount paid out in salaries). The method `sum()`, when applied to the appropriate column, can do exactly this:

In [202]:
blue_jays["Salary"].sum()

258022350

So, we find that the Blue Jays paid \$258,022,350 in player salaries during the 2025 season.

How does this compare to their World Series nemesis, the LA Dodgers?

---

#### ACTIVITY 💻: Calculate the team payroll for the LA Dodgers in 2025.

---
---

### Filtering Columns _and_ Rows

We have seen that the square bracket operator is powerful in it's ability to filter both columns and rows. However, it can only do one of these at a time. If we instead want to filter by rows and columns _simultaneously_, we need a new operator. This is where the method `df.loc[]` comes in. The general principle works as follows:

<div style="text-align: center;">
  <img src="https://raw.githubusercontent.com/ntstevens/DATSC101/1d015e0ae312ec485e0f6d3e2e657a8933cbc8b7/images/loc_example.svg" alt="mlb_logo" height = "200"/>
</div>

We can see that it combines:
* the logical statement used to filter rows, and 
* the list of column names to filter, and
* these are separated by a comma.

Previously we defined the `blue_jays` data frame by first filtering rows to include only those players whose `Team` was `TOR`, and then filtering columns to exclude `Team`. Using the `df.loc[]` method we can do these two steps all at once:

In [206]:
mlb_salaries.loc[mlb_salaries["Team"] == "TOR", ["Name", "Position", "Salary"]]

,Name,Position,Salary
1417,Adam Macko,SP,123900
1418,Addison Barger,SS,760000
1419,Alejandro Kirk,C,4600000
1420,Andres Gimenez,2B,10000000
1421,Angel Bastardo,P,760000
1422,Anthony Santander,LF,13500000
1423,Bo Bichette,SS,16500000
1424,Bowden Francis,SP,781500
1425,Braydon Fisher,RP,760000
1426,Brendon Little,RP,771800


---
---

### Advanced Filtering with Conjunctions and Disjunctions

The logical statements considered so far have been composed of a _single_ statement. But we can define a logical statement by combining _multiple_ statements with:

* _conjunctions_ (i.e., **and**), represented in Python by `&`
    * `logical_statement1 & logical_statement2` evaluates to `True` if _both of_ `logical_statement1` and `logical_statement2` are `True`. Otherwise it evaluates to `False`.
* _disjunctions_ (i.e., **or**), represented in Python by `|`
    * `logical_statement1 & logical_statement2` evaluates to `True` if _one or both of_ `logical_statement1` and `logical_statement2` are `True`. Otherwise it evaluates to `False`.

In [208]:
# AND Example
(2 > 3) & (2 < 5)

False

In [210]:
# OR Example
(2 > 3) | (2 < 5)

True

Suppose we want to investigate the salaries of Blue Jays pitchers. To do so, we need to filter the `mlb_salaries` data frame to find the Blue Jays **and** the pitchers.

We already know how to find the Blue Jays. The logical statement `mlb_salaries["Team"] == "TOR"` does that for us. We'll return to that momentarily.

Now we need to figure out how to find pitchers. If you scroll through either the `mlb_salaries` or `blue_jays` data frames, you'll see this is complicated by the fact that many players play multiple positions. So, we're looking for players whose `Position` is one of `[P, SP, RP]`. There are many ways to do this, but one elegant solution is to notice that only pitchers have a `P` in their `Position` value. So, we could filter rows so that the `Position` string contains a `P`. Fortunately the `str.contains()` method does exactly this. Let's illustrate this on the `blue_jays` data frame.

In [212]:
blue_jays["Position"].str.contains("P")

1417     True
1418    False
1419    False
1420    False
1421     True
1422    False
1423    False
1424     True
1425     True
1426     True
1427     True
1428    False
1429    False
1430     True
1431     True
1432     True
1433    False
1434    False
1435    False
1436     True
1437     True
1438     True
1439    False
1440    False
1441     True
1442     True
1443     True
1444     True
1445    False
1446     True
1447     True
1448     True
1449    False
1450    False
1451     True
1452     True
1453     True
1454     True
1455     True
1456     True
1457     True
1458     True
1459     True
1460    False
1461    False
1462    False
1463     True
1464     True
Name: Position, dtype: bool

This clearly did _something_; the result is boolean series. To see that this in fact worked as intended, let's add this column to the `blue_jays` data frame.

In [214]:
is_pitcher = blue_jays["Position"].str.contains("P")
blue_jays["Is_A_Pitcher?"] = is_pitcher
blue_jays

,Name,Position,Salary,Is_A_Pitcher?
1417,Adam Macko,SP,123900,True
1418,Addison Barger,SS,760000,False
1419,Alejandro Kirk,C,4600000,False
1420,Andres Gimenez,2B,10000000,False
1421,Angel Bastardo,P,760000,True
1422,Anthony Santander,LF,13500000,False
1423,Bo Bichette,SS,16500000,False
1424,Bowden Francis,SP,781500,True
1425,Braydon Fisher,RP,760000,True
1426,Brendon Little,RP,771800,True


Success! We can see that every pitcher has a value of `True` in the `Is_A_Pitcher?` column, and every other play has a `False`. This provides assurance that `mlb_salaries["Position"].str.contains("P")` is in fact a logical statement that can be used to filter rows to retain only pitchers.

Let's put this altogether now:

In [215]:
logical_statement1 = (mlb_salaries["Team"] == "TOR")
logical_statement2 = (mlb_salaries["Position"].str.contains("P"))
mlb_salaries[logical_statement1 & logical_statement2]

,Name,Team,Position,Salary
1417,Adam Macko,TOR,SP,123900
1421,Angel Bastardo,TOR,P,760000
1424,Bowden Francis,TOR,SP,781500
1425,Braydon Fisher,TOR,RP,760000
1426,Brendon Little,TOR,RP,771800
1427,Chris Bassitt,TOR,SP,21000000
1430,Dillon Tate,TOR,RP,1400000
1431,Easton Lucas,TOR,SP,800000
1432,Eric Lauer,TOR,RP,2200000
1436,Jacob Barnes,TOR,RP,1400000


---

#### ACTIVITY 💻: In 2025, how much more did the Blue Jays pay for pitchers than outfielders?

---
---

### Adding and Updating Columns

You may have missed it, but in the last section you saw how to _add_ a new column to a data frame. 

* Suppose you have values stored in a series called `new_data` and you want to add it the data frame `df` with the name `"My_New_Column"`. The command to do so is:

```
df["My_New_Column"] = new_data
```

You can analogously _update_ an existing column in a data frame. 

* Suppose the data frame `df` contains a column called `"Pizza"` and you want to replace its contents with the `new_data` series described above. The command to do so is:

```
df["Pizza"] = new_data
```

---
---

### Summary of `DataFrame` Methods

We have so far learned several useful methods that may be applied to `DataFrame` objects which we summarize here. Suppose `df` is a data frame.

|Method | Description |
| --- | --- |
|`df.shape` | Returns the number of rows and columns in `df` | 
|`df.column` | Returns the names of the columns in `df` | 
|`df.rename` | Allows you to rename the columns in `df`| 
|`df.drop` | Allows you to remove one or more columns in `df`| 
|`df.info()` | Provides a summary of `df`| 
|`df.sort_values()` | Sorts the rows of `df` by one or more columns| 

We also saw some helpful methods that could be applied directly to columns of `df`.

|Method | Description |
| --- | --- |
|`df["col_name"].sum()` | Sums the values contained in `"col_name"`| 
|`df["col_name"].str.contains()]` | Checks whether the values in `"col_name"` contain the given string | 


We will continue to expand these lists as we progress through the course.

---